File: javascript/src/tuple/Couple.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Set} from "@joookiwi/type"

import type {CollectionHolderOf2} from "../CollectionHolderOf2"

import {LateRetriever} from "../LateRetriever"
import {AbstractTuple} from "./AbstractTuple"

export class Couple<const T = unknown,
    const T1 extends T = T,
    const T2 extends T = T, >
    extends AbstractTuple<T> {

    //#region -------------------- Fields --------------------

    public readonly 0: T1
    public readonly 1: T2
    readonly #value1: T1
    readonly #value2: T2
    #array?: readonly [T1, T2,]
    #set?: Set<| T1 | T2>
    #collection?: CollectionHolderOf2<T, T1, T2>
    readonly #hasNull: boolean
    readonly #hasNoNulls: boolean
    readonly #hasDuplicate: boolean
    readonly #hasNoDuplicates: boolean

    //#endregion -------------------- Fields --------------------
    //#region -------------------- Constructor --------------------

    public constructor(value1: T1, value2: T2,) {
        super()
        this[0] = this.#value1 = value1
        this[1] = this.#value2 = value2
        this.#hasNoNulls = !(this.#hasNull = value1 == null || value2 == null)
        this.#hasNoDuplicates = !(this.#hasDuplicate = value1 as T === value2 as T)
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    /** The 1st value passed through the {@link constructor} held by this instance */
    public get value1(): T1 { return this.#value1 }

    /** The 2nd value passed through the {@link constructor} held by this instance */
    public get value2(): T2 { return this.#value2 }


    public override get size(): 2 { return 2 }

    public override get isEmpty(): false { return false }
    public override get isNotEmpty(): true { return true }

    public override get hasExactly1Element(): false { return false }
    public override get hasAtMost1Element(): false { return false }

    public override get hasAtLeast2Elements(): true { return true }
    public override get hasExactly2Elements(): true { return true }
    public override get hasAtMost2Elements(): true { return true }


    public override get hasNull(): boolean { return this.#hasNull }
    public override get hasNoNulls(): boolean { return this.#hasNoNulls }

    public override get hasDuplicate(): boolean { return this.#hasDuplicate }
    public override get hasNoDuplicates(): boolean { return this.#hasNoDuplicates }


    public override toArray(): readonly [T1, T2,] { return this.#array ??= Object.freeze([this.value1, this.value2,],) }

    public override toSet(): Set<| T1 | T2> { return this.#set ??= Object.freeze(new Set([this.value1, this.value2,],),) }

    public override toCollection(): CollectionHolderOf2<T, T1, T2> { return this.#collection ??= new LateRetriever.CollectionHolderOf2(this.value1, this.value2,) }

    public override toString(): string {
        const value1 = this.value1
        const value2 = this.value2
        return `(${value1 == null ? '' : value1}, ${value2 == null ? '' : value2})`
    }

    //#endregion -------------------- Methods --------------------

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/tuple/Couple.java
package joookiwi.collection.java.tuple;

import joookiwi.collection.java.CollectionHolderOf2;
import joookiwi.collection.java.method.ArrayCreator;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.MustBeInvokedByOverriders;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_NEW_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;
import static joookiwi.collection.java.method.AsString.asString;

/// An instance of [Tuple] with only 2 values from its `constructor`
///
/// @param <T>  The type
/// @param <T1> The 1<sup>st</sup> type
/// @param <T2> The 2<sup>nd</sup> type
@NotNullByDefault
public class Couple<T extends @Nullable Object,
        T1 extends T,
        T2 extends T>
        extends AbstractTuple<T>
        implements Tuple<T> {

    //#region -------------------- Fields --------------------

    private final T1 __value1;
    private final T2 __value2;

    private T @Nullable [] __array;

    //#endregion -------------------- Fields --------------------
    //#region -------------------- Constructor --------------------

    public Couple(final T1 value1, final T2 value2) {
        super();
        __value1 = value1;
        __value2 = value2;
    }

    //#endregion -------------------- Constructor --------------------
    //#region -------------------- Methods --------------------

    /// The 1<sup>st</sup> value passed through the `constructor` held by this instance
    public T1 value1() { return __value1; }

    /// The 2<sup>nd</sup> value passed through the `constructor` held by this instance
    public T2 value2() { return __value2; }

    //#region -------------------- Size methods --------------------

    @Override public @Range(from = 2, to = 2) int size() { return 2; }

    @Contract(ALWAYS_FALSE_0) @Override public boolean isEmpty() { return false; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean isNotEmpty() { return true; }

    @Contract(ALWAYS_FALSE_0) @Override public boolean hasExactly1Element() { return false; }
    @Contract(ALWAYS_FALSE_0) @Override public boolean hasAtMost1Element() { return false; }

    @Contract(ALWAYS_TRUE_0) @Override public boolean hasAtLeast2Elements() { return true; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean hasExactly2Elements() { return true; }
    @Contract(ALWAYS_TRUE_0) @Override public boolean hasAtMost2Elements() { return true; }

    //#endregion -------------------- Size methods --------------------
    //#region -------------------- Has ‥ --------------------

    @Override public boolean hasNull() { return value1() == null || value2() == null; }

    @Override public boolean hasNoNulls() { return value1() != null && value2() != null; }


    @Override public boolean hasDuplicate() {
        final var value1 = value1();
        if (value1 == null)
            return value2() == null;

        final var value2 = value2();
        if (value1 == value2)
            return true;
        return value1.equals(value2);
    }

    @Override public boolean hasNoDuplicates() {
        final var value1 = value1();
        if (value1 == null)
            return value2() != null;

        final var value2 = value2();
        if (value1 == value2)
            return false;
        return !value1.equals(value2);
    }

    //#endregion -------------------- Has ‥ --------------------
    //#region -------------------- Conversion methods --------------------

    @Override public T[] toArray() {
        final var value = __array;
        if (value != null)
            return value;
        return __array = ArrayCreator.Array(value1(), value2());
    }

    @Override public CollectionHolderOf2<T, T1, T2> toCollection() { return new CollectionHolderOf2<>(value1(), value2()); }

    @Override public String toString() { return "Couple (" + asString(value1()) + ", " + asString(value2()) + ')'; }

    //#endregion -------------------- Conversion methods --------------------
    //#region -------------------- Cloneable methods --------------------

    @MustBeInvokedByOverriders
    @Contract(ALWAYS_NEW_0)
    @Override public Couple<T, T1, T2> clone() {
        return (Couple<T, T1, T2>) super.clone();
    }

    //#endregion -------------------- Cloneable methods --------------------

    //#endregion -------------------- Methods --------------------

}
